odc/geo/cog/_mpu.py

```diff
@@ -145,93 +151,241 @@ def merge(
                 rhs.is_final,
                 lhs.lhs_keep,
             )
-
-        # Flush `lhs.data + rhs.left_data` if we can
-        #  or else move it into .left_data
         lhs.flush_rhs(write, rhs.left_data)
-
+        # Use a Dask distributed lock to ensure the nextPartId update is atomic
+        lock = Lock("mpu_merge_lock")
+        with lock:
+            all_parts = lhs.parts + rhs.parts
+            all_parts.sort(key=lambda p: int(p["PartNumber"]))
+            new_nextPartId = (
+                all_parts[-1]["PartNumber"] + 1
+                if all_parts
+                else max(lhs.nextPartId, rhs.nextPartId)
+            )
         return MPUChunk(
-            rhs.nextPartId,
+            new_nextPartId,
             rhs.write_credits,
             rhs.data,
             lhs.left_data,
-            lhs.parts + rhs.parts,
+            all_parts,
             lhs.observed + rhs.observed,
             rhs.is_final,
             lhs.lhs_keep,
         )
```

Comment:
NO, I commented on this before, parts must be ordered by construction, and this logic for next part selection is bogus.

Reply:
and no lock is needed, nor will this work in a distributed setting